binding.cc

```diff
@@ -3085,89 +3252,96 @@ typedef struct sn_crypto_stream_xchacha20_xor_state {
   uint64_t block_counter;
 } sn_crypto_stream_xchacha20_xor_state;
 
-js_value_t *
-sn_crypto_stream_xchacha20_xor_wrap_init (js_env_t *env, js_callback_info_t *info) {
-  SN_ARGV(3, crypto_stream_xchacha20_xor_wrap_init)
-
-  SN_ARGV_BUFFER_CAST(sn_crypto_stream_xchacha20_xor_state *, state, 0)
-  SN_ARGV_TYPEDARRAY(n, 1)
-  SN_ARGV_TYPEDARRAY(k, 2)
-
-  SN_THROWS(state_size != sizeof(sn_crypto_stream_xchacha20_xor_state), "state must be 'crypto_stream_xchacha20_xor_STATEBYTES' bytes")
-  SN_ASSERT_LENGTH(n_size, crypto_stream_xchacha20_NONCEBYTES, "n")
-  SN_ASSERT_LENGTH(k_size, crypto_stream_xchacha20_KEYBYTES, "k")
-
-  state->remainder = 0;
-  state->block_counter = 0;
-  memcpy(state->n, n_data, crypto_stream_xchacha20_NONCEBYTES);
-  memcpy(state->k, k_data, crypto_stream_xchacha20_KEYBYTES);
+static inline void
+sn_crypto_stream_xchacha20_xor_wrap_init(
+  js_env_t *,
+  js_receiver_t,
+  js_typedarray_span_t<> state,
+  js_typedarray_span_t<> n,
+  js_typedarray_span_t<> k
+) {
+  assert(state.size_bytes() == sizeof(sn_crypto_stream_xchacha20_xor_state));
+  assert(n.size_bytes() == crypto_stream_xchacha20_NONCEBYTES);
+  assert(k.size_bytes() == crypto_stream_xchacha20_KEYBYTES);
 
-  return NULL;
+  auto state_data = reinterpret_cast<sn_crypto_stream_xchacha20_xor_state *>(state.data());
```

Comment:
This could be improved by `libjstl` adding `js_typedarray_span_of_t<T>` similar to `js_arraybuffer_span_of_t<T>`.